### PR TITLE
WIP: Add a sample utilizing uCAM-III camera module

### PR DESCRIPTION
**Pull request by gon1332**
This PR includes a sample operating on [4D Systems uCAM-III](https://www.4dsystems.com.au/product/uCAM_III/) camera module. Currently, I haven't found anaything similar in Zephyr, so I've decided to write the drivers in the scope of a sample.

There can be a discussion on how we can support camera modules. One though is to probably, integrate camera modules into the sensor API. Or maybe define a new API?

My tasks for this PR include:

- [x] Take JPEG snapshots
- [ ] Take RAW snapshots
- [ ] Maybe shoot video, although it's not of main interest for me, for now

..and whatever comes out of discussion.

### CODE REVIEW

File: samples/camera/README.rst [hunk 1]
```diff
@@ -0,0 +1,28 @@
+.. _camera-sample:
+
+Camera Application
```
Comment (dbkinder) [inline samples/camera/README.rst]: Anticipating we may add other camera modules later, this label and title are a bit too generic.  Better would be:
```
.. _ucam-camera-sample:

uCAM-III Camera Application
###########################
```

File: samples/camera/README.rst [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+##################
+
+Overview
+********
+
+This camera application is actually a port of the 4D Systems uCAM-III camera
+module in Zephyr. It defines its own API.
+
```
Comment (dbkinder) [inline samples/camera/README.rst]: Add:
```
.. _4D Systems uCAM-III camera:
   https://www.4dsystems.com.au/product/uCAM_III/

File: samples/camera/README.rst [hunk 3]
```diff
@@ -0,0 +1,28 @@
+.. _camera-sample:
+
+Camera Application
+##################
+
+Overview
+********
+
+This camera application is actually a port of the 4D Systems uCAM-III camera
```
Comment (dbkinder) [inline samples/camera/README.rst]: Let's include a link to the uCAM website:
```suggestion
This camera application is a Zephyr port of the `4D Systems uCAM-III camera`_
```

File: samples/camera/README.rst [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+module in Zephyr. It defines its own API.
```
Comment (dbkinder) [inline samples/camera/README.rst]:
```suggestion
module. The camera API is defined in :zephyr_file:`samples/camera/src/ucam3.h`.
```

File: samples/camera/README.rst [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+Requirements
+************
+
+The demo assumes that a uCAM-III is connected to the UART_2 TX/RX of a Nucleo
+L496ZG. Also, the camera module should be powered by the 5V pin.
+
+Building and Running
+********************
+
+This sample outputs the process of interfacing with the camera module to the
```
Comment (dbkinder) [inline samples/camera/README.rst]: I'm not sure what "outputs the process of interfacing with the camera module" means?  Does it write some debugging output to the console?  (If so can you show an example here.) Does it only produce still images or a video stream? Is there a JPEG image captured somewhere?  What is the API presented by the camera module (link to that would be good)?  Is the whole API implemented in this port?

File: samples/camera/src/ucam3.h [hunk 6]
```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2019 Centaur Analytics, Inc.
```
Comment (dbkinder) [inline samples/camera/src/ucam3.h]: Is this [Centaur Analytics, Inc](https://www.linkedin.com/company/centaur-analytics-inc./)? Do we have the proper licensing rights to use this code?

**Review by dbkinder — CHANGES_REQUESTED** [review]
Comments on the docs below (and a licensing question).

**Comment by carlescufi** [comment]
There is a parallel PR proposing a camera API in https://github.com/zephyrproject-rtos/zephyr/pull/14666

**Review by dbkinder — APPROVED** [review]
Docs look OK, but can you explain where this code came from and if we have proper licensing?

File: samples/camera/src/ucam3.h (hunk 6 above)
Comment (gon1332) [reply samples/camera/src/ucam3.h]: Sorry for the big delay. This is the company. However I own this code, so it's now under my name.

**Comment by gon1332** [comment]
> Docs look OK, but can you explain where this code came from and if we have proper licensing?

I am the original author, and I want to offer it with the default Zephyr licence.

**Comment by gon1332** [comment]
> There is a parallel PR proposing a camera API in #14666

I've seen it. This PR is about interfacing with the 4d systems specific module. When I'm done, I'd be happy to offer an implementation to the proposing V4Zephyr API for this module.

**Comment by JunYangNXP** [comment]
@gon1332 
As the quote from https://4dsystems.com.au/ucam-iii:
"The module uses a CMOS VGA colour sensor along with a JPEG compression chip."
1) What is the "CMOS VGA colour sensor"? Any information about this to provide?
2) What is the "JPEG compression chip"?
3) The Zephy is supposed to be running on your product, right? Or on host?

**Comment by gon1332** [comment]
> As the quote from https://4dsystems.com.au/ucam-iii:
> "The module uses a CMOS VGA colour sensor along with a JPEG compression chip."
> 
> 3. The Zephy is supposed to be running on your product, right? Or on host?

@JunYangNXP Zephyr is supposed to run on host, utilizing the camera module (which, BTW is a not our product).

> 1. What is the "CMOS VGA colour sensor"? Any information about this to provide?
> 2. What is the "JPEG compression chip"?
As I do not have experience with camera modules implementations, along with lack of technical papers from 4DSystems I cannot answer these.

**Comment by loicpoulain** [comment]
@gon1332 , now that video 4 zephyr has been merged I wonder If you get any plan and have any time to adapt this driver.

**Comment by gon1332** [comment]
> @gon1332 , now that video 4 zephyr has been merged I wonder If you get any plan and have any time to adapt this driver.

I'll give it a try from this week.

**Pull request closed without merging**